### PR TITLE
Repurpose numpy-compiler compiler/runtime flow for PyTorch.

### PR DESCRIPTION
**Pull request by stellaraccident**
* A bit gross because I took the chance to upgrade all of the backend bits to the new MLIR Python bindings and we still co-mingle the old and new for now.
* Since the Python created PassManagers are configured for explicit nesting, I had to upgrade some of the pass pipelines to be explicit.
* The demo in mul_maximum_e2e.py now compiles, runs through PyTorch and through the JIT, prints and asserts the same results.
* I am not claiming that this is the prettiest API in this patch: consider that this is just directly using low-level APIs and there should be an intervening high level API.
* Requires upstream https://reviews.llvm.org/D91232

### CODE REVIEW

File: lib/RefBackend/RefBackend.cpp [hunk 1]
```diff
@@ -194,17 +194,17 @@ void mlir::NPCOMP::createRefBackendLoweringPipeline(
   //
   // Also, converting to linalg herevopens up a lot of optimization
   // opportunities.
-  pm.addPass(createConvertElementwiseToLinalgPass());
+  pm.addNestedPass<FuncOp>(createConvertElementwiseToLinalgPass());
```
Comment (silvasean) [inline lib/RefBackend/RefBackend.cpp:197]: Thanks for making this explicit! Could you try giving a stab at removing OpPassManager::Nesting::Implicit from our codebase so we don't regress? I think this was the only offender. (if that doesn't "just work", then don't worry about it)

File: lib/RefBackend/RefBackend.cpp [hunk 2]
```diff
@@ -305,6 +305,7 @@ void mlir::NPCOMP::createTCFRefBackendLoweringPipeline(
   // case of invalid broadcasts.
   //
   // TCP does not. So we need to reify the broadcasting and error checking.
+  // These all run at the module level.
```
Comment (silvasean) [inline lib/RefBackend/RefBackend.cpp:308]: ugh: I'll fix that.

File: frontends/pytorch/examples/mul_maximum_e2e.py [hunk 3]
```diff
@@ -17,8 +24,13 @@
...
-#   npcomp-opt -aten-recognize-kernels -convert-aten-to-tcf \
-#     -numpy-public-functions-to-tensor -canonicalize
-mb.module.operation.print()
+backend = CompilerBackend()
+jit_module = backend.load(backend.compile(mb.module))
+
+jit_result = jit_module.mul_maximum(lhs.numpy(), rhs.numpy(), threshold.numpy(),
+                                    bias.numpy())
+
+print(f"PyTorch Result = {result.numpy()}", file=sys.stderr)
```
Comment (silvasean) [inline frontends/pytorch/examples/mul_maximum_e2e.py:33]: Wooooooooooo!!!

**Review by silvasean — APPROVED** [review]
Wooooo!!

File: lib/RefBackend/RefBackend.cpp (hunk 1 above)
Comment (stellaraccident) [reply lib/RefBackend/RefBackend.cpp:197]: I think this will be done by https://reviews.llvm.org/D91233

**Pull request merged**